src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.Http3.cs

```diff
@@ -100,103 +114,523 @@ internal sealed partial class HttpConnectionPool
+                        {
+                            Debug.Assert((_availableHttp3Connections?.Count ?? 0) == 0, $"With {(_availableHttp3Connections?.Count ?? 0)} available HTTP3 connections, we shouldn't have a waiter.");
+                        }
+                        else if (_disposed)
+                        {
+                            // The pool has been disposed. We will dispose this connection below outside the lock.
+                            // We do this check after processing the request queue so that any queued requests will be handled by existing connections if possible.
+                            _associatedHttp3ConnectionCount--;
+                        }
+                        else
+                        {
+                            // Add connection to the pool.
+                            added = true;
+                            _availableHttp3Connections ??= new List<Http3Connection>();
+                            _availableHttp3Connections.Add(connection);
+                        }
                     }
 
-                    return _http3Connection;
+                    if (waiter is not null)
+                    {
+                        Debug.Assert(!added);
+
+                        if (waiter.TrySignal(connection))
+                        {
+                            break;
+                        }
+
+                        // Loop and process the queue again
+                    }
+                    else
+                    {
+                        if (reserved)
+                        {
+                            connection.ReleaseStream();
+                        }
+                        if (added)
+                        {
+                            if (NetEventSource.Log.IsEnabled()) connection.Trace("Put HTTP3 connection in pool.");
+                            return;
+                        }
+                        else
+                        {
+                            Debug.Assert(_disposed);
+                            if (NetEventSource.Log.IsEnabled()) connection.Trace("Disposing HTTP3 connection returned to pool. Pool was disposed.");
+                            connection.Dispose();
+                            return;
+                        }
+                    }
                 }
+            }
+
+            if (isNewConnection)
+            {
+                Debug.Assert(initialRequestWaiter is not null, "Expect request for a new connection");
 
-                if (NetEventSource.Log.IsEnabled())
+                // The new connection could not handle even one request, either because it shut down before we could use it for any requests,
+                // or because it immediately set the max concurrent streams limit to 0.
+                // We don't want to get stuck in a loop where we keep trying to create new connections for the same request.
+                // So, treat this as a connection failure.
```

Comment:
I don't think this is right for HTTP/3.

With HTTP/2, this would only happen if the connection was already closed as it otherwise always starts with a non-zero amount of available streams.

With HTTP/3, `_availableRequestStreamsCount` may still be 0 at this point, especially given that `StreamsAvailableCallback` is queued to the thread pool and may therefore signal the initial available streams _after_ we hand the connection to the pool and reach this block.

Reply:
You're right, fixed.